File: 1kyu/tiny_three_pass_compiler.js
function Token(tokens) {
  let ndx = 0;
  return new class {
    getToken() {
      return tokens[ndx++];
    }
    ungetToken() {
      ndx--;
    }
  };
}

function Compiler() {
  const args = [];

  function parseFunction(tokens) {
    tokens.getToken(); // pass [
    let token = tokens.getToken();
    while (token !== ']') {
      args.push(token);
      token = tokens.getToken();
    }
  }

  function parseFactor(tokens) {
    const token = tokens.getToken();
    let v;
    if (typeof token === 'number') {
      v = { op: 'imm', n: token };
    } else if (token.match(/[a-zA-Z]/)) {
      const ndx = args.indexOf(token);
      v = { op: 'arg', n: ndx };
    } else if (token === '(') {
      v = parseExpression(tokens);
      tokens.getToken(); // pass ')'
    } else {
      tokens.ungetToken();
    }
    return v;
  }

  function parseTerm(tokens) {
    let v1 = parseFactor(tokens);
    while (true) {
      const token = tokens.getToken();
      if (token !== '*' && token !== '/') {
        tokens.ungetToken();
        break;
      } else {
        const v2 = parseFactor(tokens);
        if (token === '*') {
          v1 = { op: '*', a: v1, b: v2 };
        } else if (token === '/') {
          v1 = { op: '/', a: v1, b: v2 };
        } else {
          tokens.ungetToken();
        }
      }
    }
    return v1;
  }

  function parseExpression(tokens) {
    let v1 = parseTerm(tokens);
    while (true) {
      const token = tokens.getToken();
      if (token !== '-' && token !== '+') {
        tokens.ungetToken();
        break;
      } else {
        const v2 = parseTerm(tokens);
        if (token === '+') {
          v1 = { op: '+', a: v1, b: v2 };
        } else if (token === '-') {
          v1 = { op: '-', a: v1, b: v2 };
        } else {
          tokens.ungetToken();
        }
      }
    }
    return v1;
  }

  function reduceConstant(ast) {
    const { op } = ast;
    if (op === 'imm' || op === 'arg') {
      return ast;
    }
    const a = reduceConstant(ast.a);
    const b = reduceConstant(ast.b);
    if (a.op === 'imm' && b.op === 'imm') {
      let n;
      switch (op) {
        case '+':
          n = a.n + b.n;
          break;
        case '-':
          n = b.n - a.n;
          break;
        case '*':
          n = a.n * b.n;
          break;
        case '/':
          n = a.n / b.n;
          break;
        default:
          break;
      }
      return { op: 'imm', n };
    }
    return { op, a, b };
  }

  const ops = { '+': 'AD', '-': 'SU', '*': 'MU', '/': 'DI' };
  function assembly(ast) {
    const { op } = ast;
    if (op === 'imm') {
      return [`IM ${ast.n}`];
    } else if (op === 'arg') {
      return [`AR ${ast.n}`];
    }
    const a = assembly(ast.a);
    const b = assembly(ast.b);
    if (a.length === 1 && b.length === 1) {
      if (op === '-') {
        return [b[0], 'SW', a[0], ops[op]];
      }
      return [a[0], 'SW', b[0], ops[op]];
    }

    a.push('PU');
    b.push('SW');
    b.push('PO');
    b.push(ops[op]);
    return a.concat(b);
  }

  return new class {
    compile(program) {
      return this.pass3(this.pass2(this.pass1(program)));
    }
    tokenize(program) {
      // Turn a program string into an array of tokens.  Each token
      // is either '[', ']', '(', ')', '+', '-', '*', '/', a variable
      // name or a number (as a string)
      const regex = /\s*([-+*/\(\)\[\]]|[A-Za-z]+|[0-9]+)\s*/g;
      return program.replace(regex, ':$1')
        .substring(1)
        .split(':')
        .map(tok => isNaN(tok) ? tok : tok | 0);
    }

    pass1(program) {
      const tokens = new Token(this.tokenize(program));
      parseFunction(tokens);
      return parseExpression(tokens);
    }

    pass2(ast) {
      return reduceConstant(ast);
    }
    pass3(ast) {
      console.log(JSON.stringify(ast));
      return assembly(ast);
    }
  };
}

function simulate(asm, args) {
  let r0 = undefined;
  let r1 = undefined;
  const stack = [];
  asm.forEach(instruct => {
    const match = instruct.match(/(IM|AR)\s+(\d+)/) || [0, instruct, 0];
    const ins = match[1];
    const n = match[2] | 0;

    if (ins === 'IM') {
      r0 = n;
    } else if (ins === 'AR') {
      r0 = args[n];
    } else if (ins === 'SW') {
      const tmp = r0; r0 = r1; r1 = tmp;
    } else if (ins === 'PU') {
      stack.push(r0);
    } else if (ins === 'PO') {
      r0 = stack.pop();
    } else if (ins === 'AD') {
      r0 += r1;
    } else if (ins === 'SU') {
      r0 -= r1;
    } else if (ins === 'MU') {
      r0 *= r1;
    } else if (ins === 'DI') {
      r0 /= r1;
    }
  });
  return r0;
}

const prog = '[ x y z ] x - y - z + 10 / 5 / 2 - 7 / 1 / 7';

const c = new Compiler();

const ins = c.compile(prog);
console.log(JSON.stringify(ins));
console.log(simulate(ins, [5,4,1]));

File: 1kyu/tiny_three_pass_compiler.dart
import "package:solution/preloaded.dart";

class Compiler {
  List<String> args = [];
  List<String> tokens = [];
  int ndx = 0;

  String getToken() {
    return tokens[ndx++];
  }
  
  void ungetToken() {
    ndx--;
  }

  List<String> tokenize(String prog) {
    List<String> tokens = [];
    RegExp pattern = RegExp("[-+*/()\\[\\]]|[a-zA-Z]+|\\d+");
    pattern.allMatches(prog).toList().forEach((m) {
        tokens.add(m.group(0));
    });
    tokens.add("\$"); // end-of-stream
    return tokens;
  }

  List<String> compile(String prog) => pass3(pass2(pass1(prog)));

  bool isDigit(int s) => s >= '0'.codeUnitAt(0) && s <= '9'.codeUnitAt(0);
  bool isAlpha(int s) => s >= 'a'.codeUnitAt(0) && s <= 'z'.codeUnitAt(0)
  || s >= 'A'.codeUnitAt(0) && s <= 'Z'.codeUnitAt(0);
  
  void parseFunction() {
    getToken();
    var token = getToken();
    while (token != ']') {
      args.add(token);
      token = getToken();
    }
  }

  Ast parseFactor() {
    var token = getToken();
    Ast v;
    if (isDigit(token.codeUnitAt(0))) {
      v = UnOp('imm', int.parse(token));
    } else if (isAlpha(token.codeUnitAt(0))) {
      int n = args.indexOf(token);
      v = UnOp('arg', n);
    } else if (token == '(') {
      v = parseExpression();
      getToken();
    } else {
      ungetToken();
    }
    return v;
  }

  Ast parseExpression() {
    Ast v1 = parseTerm();
    while (true) {
      var token = getToken();
      if (token != '-' && token != '+') {
        ungetToken();
        break;
      } else {
        Ast v2 = parseTerm();
        if (token == '+') {
          v1 = BinOp('+', v1, v2);
        } else if (token == '-') {
          v1 = BinOp('-', v1, v2);
        } else {
          ungetToken();
        }
      }
    }
    return v1;
  }

  Ast parseTerm() {
    Ast v1 = parseFactor();
    while (true) {
      var token = getToken();
      if (token != '*' && token != '/') {
        ungetToken();
        break;
      } else {
        Ast v2 = parseFactor();
        if (token == '*') {
          v1 = BinOp('*', v1, v2);
        } else if (token == '/') {
          v1 = BinOp('/', v1, v2);
        } else {
          ungetToken();
        }
      }
    }
    return v1;
  }

  Ast reduceConstant(Ast ast) {
    var op = ast.op();
    if (op == 'imm' || op == 'arg') {
      return ast;
    }
    BinOp binOp = ast;;
    Ast a1 = reduceConstant(binOp.a);
    Ast b1 = reduceConstant(binOp.b);
    if (a1.op() == 'imm' && b1.op() == 'imm') {
      int n;
      UnOp a = a1;
      UnOp b = b1;
      if (op == '+') {
        n = a.n + b.n;
      } else if (op == '-') {
        n = a.n - b.n;
      } else if (op == '*') {
        n = a.n * b.n;
      } else {
        n = a.n ~/ b.n;
      }
      return UnOp('imm', n);
    }

    return BinOp(op, a1, b1);
  }

  List<String> assembly(Ast ast) {
    String op = ast.op();
    if (op == 'imm') {
      return ['IM ${(ast as UnOp).n}'];
    }
    if (op == 'arg') {
      return ['AR ${(ast as UnOp).n}'];
    }
    List<String> a = assembly((ast as BinOp).a);
    List<String> b = assembly((ast as BinOp).b);
    Map<String, String> ops = { '+': 'AD', '-': 'SU', '*': 'MU', '/': 'DI' };
    if (a.length == 1 && b.length == 1) {
      if (op == '-') {
        return [b[0], 'SW', a[0], ops[op]];
      }
      return [a[0], 'SW', b[0], ops[op]];
    }

    a.add('PU');
    b.add('SW');
    b.add('PO');
    b.add(ops[op]);
    return a + b;
  }
  /**
  * Returns an un-optimized AST
  */
  Ast pass1(String prog) {
    tokens = tokenize(prog);
    parseFunction();
    return parseExpression();
  }

  /**
  * Returns an AST with constant expressions reduced
  */
  Ast pass2(Ast ast) {
    return reduceConstant(ast);
  }

  List<String> pass3(Ast ast) {
    return assembly(ast);
  }
}
